crates/sui-framework/deps/move-stdlib/sources/u64.move

```diff
@@ -0,0 +1,170 @@
+        if (x < y) {
+            x
+        } else {
+            y
+        }
+    }
+
+    /// Return the value of a base raised to a power.
+    /// This function is not overflow-safe and will abort with an overflow error if `base^exponent` > `MAX`
+    public fun pow(base: u64, exponent: u8): u64 {
+        let res = 1;
+        while (exponent >= 1) {
+            if (exponent % 2 == 0) {
+                base = base * base;
+                exponent = exponent / 2;
+            } else {
+                res = res * base;
+                exponent = exponent - 1;
+            }
+        };
+
+        res
+    }
+
+    /// Return the mean of x and y, with the remainder truncated.
+    /// This function is overflow-safe and cannot abort
+    public fun mean(x: u64, y: u64): u64 {
+        if (x < y) {
+            x + (y - x) / 2
+        } else {
+            y + (x - y) / 2
+        }
+    }
+
+    /// Get a nearest lower integer square root for `x`. Given that this
+    /// function can only operate with integers, it is impossible
+    /// to get perfect (or precise) integer square root for some numbers.
+    ///
+    /// Example:
+    /// ```
+    /// math::sqrt(9) => 3
+    /// math::sqrt(8) => 2 // the nearest lower square root is 4;
+    /// ```
+    ///
+    /// In integer math, one of the possible ways to get results with more
+    /// precision is to use higher values or temporarily multiply the
+    /// value by some bigger number. Ideally if this is a square of 10 or 100.
+    ///
+    /// Example:
+    /// ```
+    /// math::sqrt(8) => 2;
+    /// math::sqrt(8 * 10000) => 282;
+    /// // now we can use this value as if it was 2.82;
+    /// // but to get the actual result, this value needs
+    /// // to be divided by 100 (because sqrt(10000)).
+    ///
+    ///
+    /// math::sqrt(8 * 1000000) => 2828; // same as above, 2828 / 1000 (2.828)
+    /// ```
+    public fun sqrt(x: u64): u64 {
+        let bit = 1u128 << 64;
+        let res = 0u128;
+        let x = (x as u128);
+
+        while (bit != 0) {
+            if (x >= res + bit) {
+                x = x - (res + bit);
+                res = (res >> 1) + bit;
+            } else {
+                res = res >> 1;
+            };
+            bit = bit >> 2;
+        };
+
+        (res as u64)
+    }
+
+    /// Like the + operator, but returns a None on overflow instead of aborting
+    public fun safe_add(x: u64, y: u64): Option<u64> {
+        if (y > MAX - x) {
+            option::none()
+        } else {
+            option::some(x + y)
+        }
+    }
+
+    /// Like the + operator, but allows overflow rather instead of aborting
+    public fun saturating_add(x: u64, y: u64): u64 {
+        abort(0) // implement me
+    }
+
+    /// Compute the sum of all the values in `nums`
+    /// This function is not overflow-safe and will abort with an overflow error if the sum > `MAX`
+    public fun vec_sum(nums: &vector<u64>): u64 {
```

Comment:
Not sure about having these operations in here -- curious to hear other people's thoughts -- I think I'd prefer us investing in easier to use looping primitives for Move.

Reply:
I think putting `vec_sum`, `vec_mean`, etc elsewhere makes sense. Curious for where folks thing `vec_sort` should live, though (here? A separate `sort` module that exposes sorting for all ground types)?

Reply:
Too much to all put in `vector`? 

The only convenient thing about having them here in `u64` is you can keep the name `vec_sum`. Maybe it's just reasonable just having the name `sum`? `u64::sum`

Reply:
My thoughts is you will probably want a sum/mean for all the integer types. And keeping them in their own module feels reasonable?